FF1Lib/RewardSourcePicker.cs

```diff
@@ -0,0 +1,104 @@
+﻿using RomUtilities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace FF1Lib
+{
+	public class RewardSourcePicker
+	{
+		Dictionary<int, double> weights = new Dictionary<int, double>();
+
+		double reduction;
+		double nonchest;
+
+		ISanityChecker checker;
+
+		public RewardSourcePicker(double _reduction, double _nonchest, ISanityChecker _checker)
+		{
+			reduction = _reduction;
+			nonchest = _nonchest;
+			checker = _checker;
+		}
+
+		public IRewardSource Pick(List<IRewardSource> sources, bool forward, bool spread, bool incentive, MT19337 rng)
+		{
+			if (!sources.Any()) return null;
+
+			IRewardSource result = null;
+
+			//This loop sums up all weights of the reward source.
+			double sum = 0.0;
+			foreach (var s in sources)
+			{
+				//If weight already exists for rewardsource use it
+				if (weights.TryGetValue(s.Address, out var v))
+				{
+					sum += v;
+				}
+				//If it doesn't have a weight and it's a chest, give it the weight 1
+				else if (s is TreasureChest)
+				{
+					sum += 1.0;
+					weights.Add(s.Address, 1.0);
+				}
+				//If it's not a chest, give it a significantly higher weight. There are way more chests than nonchet rewards.
+				//That way there is a reasonable chance, that some loose key items land on npcs.
+				else if (!incentive)
+				{
+					sum += nonchest;
+					weights.Add(s.Address, nonchest);
+				}
+				else
+				{
+					sum += 1.0;
+					weights.Add(s.Address, nonchest);
+				}
+			}
+
+			//We pick a value between 0 and sum. We can't work with the full double range, but with reasonable weight reduction, it's not going to be a problem.
+			var r = rng.Next() / (double)uint.MaxValue * sum;
+
+			//This loop finds the reward source associated with the rng number.
+			//If Forward placement is active, reduce the weight of all reward sources that were elligible for an item this step.
+			//Normally chest like ToF or Matoya have a chance to roll a loose item at every step, because they are accessible from the start.
+			//TFC is accessible way later so is used in way less rolls.
+			//By reducing the weight on chests, that already had a chance, it's more likely to select a chest or npc from a newly opened up area(hence forward placement).
+			//The forward placement is deactivated for incentive items, so it doesn't skew the placement.
+			sum = 0.0;
+			foreach (var s in sources)
+			{
+				var v = weights[s.Address];
+				if (forward) weights[s.Address] = v * reduction;
```

Comment:
Since you're multiplying by 50% repeatedly you should set a numeric floor on weights, e.g. Math.Min(v * reduction, .0000000001)

Same for spread

Reply:
The limit is kind of useless. When a chest falls below 1 in 4 billions, it may as well drop out completely.
All chests dropping to low values is well covered by double bevavior.
Should any numerical issues really arise, there is a safety net that doesn't let if fail.

Reply:
ok